### PR TITLE
[Feature] Support more MMOCR models: DBNet++, TextSnake, ABINet, MRCNN

### PR DESCRIPTION
**Pull request by AllentDan**
To support more MMOCR models.
- [x] MRCNN
- [x] DBNet++
- [x] TextSnake
- [x] ABINet

Lacks:
- [x] Unit Tests
- [x] Docs && Benchmark

### CODE REVIEW

**Comment by lvhan028** [comment]
Does SDK support them?

**Comment by AllentDan** [comment]
> Does SDK support them?

Not all of them were supported.

File: configs/mmocr/text-detection/text-detection_mrcnn_tensorrt-fp16_dynamic-320x320-2240x2240.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
+_base_ = [
```
Comment (grimoire) [inline configs/mmocr/text-detection/text-detection_mrcnn_tensorrt-fp16_dynamic-320x320-2240x2240.py:1]: if we base `text-detection_mrcnn_tensorrt`, we just need to enable fp16 here.

File: docs/en/03-benchmark/benchmark.md [hunk 2]
```diff
@@ -1143,6 +1143,42 @@ Users can directly test the performance through [how_to_evaluate_a_model.md](../
...
+    <td align="center" rowspan="3">TextDetection</td>
+    <td align="center" rowspan="3">ICDAR2015</td>
+    <td align="center">recall</td>
+    <td align="center">0.8209</td>
+    <td align="center">0.8209</td>
+    <td align="center">0.8209</td>
+    <td align="center">0.8199</td>
+    <td align="center">0.8204</td>
+    <td align="center">0.8204</td>
+    <td>-</td>
```
Comment (grimoire) [inline docs/en/03-benchmark/benchmark.md:1157]: align?

File: docs/zh_cn/03-benchmark/benchmark.md [hunk 3]
```diff
@@ -1138,6 +1138,42 @@ GPU: ncnn, TensorRT, PPLNN
...
+    <td align="center" rowspan="3">TextDetection</td>
+    <td align="center" rowspan="3">ICDAR2015</td>
+    <td align="center">recall</td>
+    <td align="center">0.8209</td>
+    <td align="center">0.8209</td>
+    <td align="center">0.8209</td>
+    <td align="center">0.8199</td>
+    <td align="center">0.8204</td>
+    <td align="center">0.8204</td>
+    <td>-</td>
```
Comment (grimoire) [inline docs/zh_cn/03-benchmark/benchmark.md:1152]: align

File: mmdeploy/codebase/base/task.py [hunk 4]
```diff
@@ -111,6 +111,8 @@ def build_pytorch_model(self,
         preprocess_cfg.update(
             deepcopy(self.model_cfg.get('data_preprocessor', {})))
         model.setdefault('data_preprocessor', preprocess_cfg)
+        if model.type == 'MMDetWrapper':  # Mask-RCNN in MMOCR
```
Comment (grimoire) [inline mmdeploy/codebase/base/task.py:114]: override build_pytorch_model instead of update in base task

File: mmdeploy/codebase/mmocr/models/text_recognition/abi_language_decoder.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    """
+    # out as a boolean vector indicating the existence of end token(s)
+    out = (logit.argmax(dim=-1) == self.dictionary.end_idx)
+    abn = out.any(dim)
+    # Get the first index of end token
+    # add `.float()` to `out` for onnxruntime `cumsum()`
+    # add `.float()` before `argmax()`
+    out = ((out.float().cumsum(dim) == 1) & out).float().argmax(dim)
+    out = out + 1
+    out = torch.where(abn, out, out.clone().detach()).float()
```
Comment (grimoire) [inline mmdeploy/codebase/mmocr/models/text_recognition/abi_language_decoder.py:29]: What's this `where` for?

File: mmdeploy/codebase/mmocr/models/text_recognition/transformer_module.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+        ]).to(device)
+        denominator = denominator.view(1, -1)
+        pos_tensor = torch.arange(n_position).to(device).unsqueeze(-1).float()
+        sinusoid_table = pos_tensor * denominator
+        sinusoid_table[:, 0::2] = torch.sin(sinusoid_table[:, 0::2])
+        sinusoid_table[:, 1::2] = torch.cos(sinusoid_table[:, 1::2])
+
+        return sinusoid_table.unsqueeze(0)
+
+    def forward(self, x):
```
Comment (grimoire) [inline mmdeploy/codebase/mmocr/models/text_recognition/transformer_module.py:35]: Why do we need a module rewriter here?

File: mmdeploy/pytorch/functions/any.py [hunk 7]
```diff
@@ -0,0 +1,24 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import torch
+
+from mmdeploy.core import FUNCTION_REWRITER
+
+
+@FUNCTION_REWRITER.register_rewriter(func_name='torch.Tensor.any')
+@FUNCTION_REWRITER.register_rewriter(func_name='torch.any')
+def any__default(input, *args, **kwargs) -> torch.Tensor:
```
Comment (grimoire) [inline mmdeploy/pytorch/functions/any.py:9]: add dim and keepdim as parameters with default value here. so you don't have to check these params below.

File: mmdeploy/codebase/mmocr/models/text_recognition/abi_language_decoder.py (hunk 5 above)
Comment (AllentDan) [reply mmdeploy/codebase/mmocr/models/text_recognition/abi_language_decoder.py:29]: Just copied from MMOCR [here](https://github.com/open-mmlab/mmocr/blob/ebdf1cf90d44e278a322f327ac18eac6fff5a745/mmocr/models/textrecog/decoders/abi_language_decoder.py#L194)

File: mmdeploy/codebase/mmocr/models/text_recognition/transformer_module.py (hunk 6 above)
Comment (AllentDan) [reply mmdeploy/codebase/mmocr/models/text_recognition/transformer_module.py:35]: [Original implementation](https://github.com/open-mmlab/mmocr/blob/ebdf1cf90d44e278a322f327ac18eac6fff5a745/mmocr/models/common/modules/transformer_module.py#L130) would raise device error when we use cuda.

File: mmdeploy/codebase/mmocr/models/text_recognition/abi_language_decoder.py (hunk 5 above)
Comment (grimoire) [reply mmdeploy/codebase/mmocr/models/text_recognition/abi_language_decoder.py:29]: Did `out.new_tensor(logit.shape[1])==out`?
What if it did, `torch.where(abn, out, out.clone().detach())==out` whatever abn is, right?

File: mmdeploy/codebase/mmocr/models/text_recognition/transformer_module.py (hunk 6 above)
Comment (grimoire) [reply mmdeploy/codebase/mmocr/models/text_recognition/transformer_module.py:35]: Why not rewrite the forward instead of the module?

Comment (AllentDan) [reply mmdeploy/codebase/mmocr/models/text_recognition/transformer_module.py:35]: the tensor was registered in `init` method

**Review by grimoire — APPROVED** [review]
LGTM

**Review by lzhangzz — APPROVED** [review]
